### PR TITLE
Adding logdet

### PR DESCRIPTION
**Pull request by bhargavvader**
This is with regard to [#1012](https://github.com/pymc-devs/pymc3/issues/1012). 
The code is copy pasted from Theano PR [#3959](https://github.com/Theano/Theano/pull/3959) as suggested in the issue, with the function renamed to `logdet`.

Should there be tests made for this as well?

### CODE REVIEW

**Comment by twiecki** [comment]
Thanks @bhargavvader! Can we just copy the test from the PR?

File: pymc3/theanof.py [hunk 1]
```diff
@@ -59,6 +60,37 @@ def floatX(X):
 Theano derivative functions
 """
 
+class logdet(Op):
+    """Computes the logarithm of absolute determinant of a square
+    matrix M, log(abs(det(M))), on CPU. Avoids det(M) overflow/
+    underflow.
+
+    Note: Once PR #3959 in the Theano repo is merged, this must be removed.
```
Comment (twiecki) [inline pymc3/theanof.py]: Should add specific link to PR and also mention the original author.

**Comment by twiecki** [comment]
Moreover, would be great if this functionality would already be used in ADVI as part of this PR.

File: pymc3/tests/test_theanof.py [hunk 2]
```diff
@@ -18,6 +19,33 @@ def integers_ndim(ndim):
         yield np.ones((2,) * ndim) * i
         i += 1
 
+class TestLogDet(unittest.TestCase):
```
Comment (twiecki) [inline pymc3/tests/test_theanof.py]: Should inherit from SeededTest

**Comment by bhargavvader** [comment]
Addressed code comments. 
What changes to ADVI would be needed? Is it in the `ELBO` part of the code?

File: pymc3/theanof.py [hunk 3]
```diff
@@ -1,10 +1,12 @@
...
 from .vartypes import typefilter, continuous_types
 from theano import theano, scalar, tensor as tt
 from theano.gof.graph import inputs
-from theano.gof import Op
+from theano.gof import Op, Apply
 from theano.configparser import change_flags
+from theano.tensor.nlinalg import matrix_inverse
 from .memoize import memoize
 from .blocking import ArrayOrdering
 from .data import DataGenerator
```
Comment (twiecki) [inline pymc3/theanof.py]: Need to add to `__all__`, forgot about that.

**Comment by twiecki** [comment]
https://github.com/pymc-devs/pymc3/blob/master/pymc3/distributions/multivariate.py#L116 here is where we want to use it. Was wrong about ADVI.

File: pymc3/theanof.py (hunk 3 above)
Comment (bhargavvader) [inline pymc3/theanof.py]: What exactly would you want to add to `__all__` ? Do you mean adding `LogDet` to it?

**Comment by bhargavvader** [comment]
Cool - so would that be doing `result = k * tt.log(2 * np.pi) - logdet(tau)` ?
Edit: here, `tt.log(1. / det(tau))` = `- logdet(tau)`, right?

Also, the tests are failing, but I think it's related to #1773. The `theanof` tests seem to be fine.

**Comment by twiecki** [comment]
> Edit: here, tt.log(1. / det(tau)) = - logdet(tau), right?

Yes, that looks right.

File: pymc3/theanof.py (hunk 3 above)
Comment (twiecki) [inline pymc3/theanof.py]: Yes, exactly, and `logdet`.

**Comment by bhargavvader** [comment]
@twiecki , made the changes to `multivariate` and added `logdet` to `__all__`. Could you review it?

File: pymc3/distributions/multivariate.py [hunk 4]
```diff
@@ -12,6 +12,8 @@
 from theano.tensor.nlinalg import det, matrix_inverse, trace
 
 import pymc3 as pm
+
+from pycm3.theanof import logdet
```
Comment (twiecki) [inline pymc3/distributions/multivariate.py]: typo pymc3

**Comment by twiecki** [comment]
@bhargavvader Make sure that unittests (specifically the distribution ones) still pass locally.

**Comment by bhargavvader** [comment]
I've corrected the typo - had a few doubts about running the tests locally, working on it now.

**Comment by bhargavvader** [comment]
All tests pass :)

**Comment by twiecki** [comment]
This is great. I want a second set of eyes on this: @ColCarroll @taku-y @fonnesbeck do you have time to take a look?

**Comment by bhargavvader** [comment]
Update: when I ran `nosetests pymc3/tests`, there were 9 errors. Both `test_distributions.py` and `test_theanof.py` failed with `ImportError: No module named nose_parameterized`. The other fails were with `pymc3.tests.test_stats.TestStats` which I reckon is unrelated. 

~~Edit: is it possible to run only certain files using nosetest?~~ 
Figured it out.

**Comment by ColCarroll** [comment]
A recent PR added nose-parametrized.  You can pip install it.

**Comment by bhargavvader** [comment]
All of `test_distributions.py`, `test_distributions_timeseries.py` and `test_distributions_random.py` pass with no errors. (I don't think these run on travis?)

**Comment by twiecki** [comment]
What makes you say they don't run on travis?

Another thing occurred to me: missing GPU support. We're inducing a blocker to run any model with MvNormal on the GPU. I see two solutions:
* Add GPU support to logdet (not sure how big of a task that would be)
* Add a flag to MvNormal to either use the numerically stable logdet or the manual version.

**Comment by bhargavvader** [comment]
Oh my bad - I noticed that one of the travis checks doesn't include them, but the other one does. 

About GPU support, I would be up to the task, but not immediately - I am caught up with some school work and can't devote too much time at the moment. We can add a flag for the moment if that makes sense. Would it be a log, or a warning?

**Comment by fonnesbeck** [comment]
I'm happy with merging this. Perhaps GPU support can be added after the Theano PR is merged?

**Comment by twiecki** [comment]
Here's what I think we should do:
* Add a flag to MvNormal: `gpu_compat=False` with a doc-string explaining the behavior
* In `__init__` check if gpu_compat is False *and* theano.config.device == 'gpu' and raise a warning alerting of the incompatibility and notifying about the gpu_compat flag at cost of numerical accuracy.

**Comment by fonnesbeck** [comment]
@twiecki Is that something we think we'd use again? If so, maybe write a decorator that could be added and removed whenever we have (temporary) GPU compatibility issues.

**Comment by twiecki** [comment]
@fonnesbeck Good point. We could put that logic into the Op itself maybe. We might also want to automatically switch in case of GPU and give a warning.

**Comment by taku-y** [comment]
I looked over the code and it looks good to me. It would be nice to merge this PR even if GPU has not been supported yet.

**Comment by bhargavvader** [comment]
Cool - I've added the flag and the warning. I agree with the idea of having a decorator for GPU compatibility issues but I'd rather take it up in a new PR (if we do want it).

File: pymc3/distributions/multivariate.py [hunk 5]
```diff
@@ -113,7 +123,7 @@ def logp(self, value):
         delta = value - mu
         k = tau.shape[0]
 
-        result = k * tt.log(2 * np.pi) + tt.log(1. / det(tau))
+        result = k * tt.log(2 * np.pi) - logdet(tau)
```
Comment (twiecki) [inline pymc3/distributions/multivariate.py]: Need to actually check if gpu_compat is true and use the previous statement.

Comment (bhargavvader) [reply pymc3/distributions/multivariate.py]: Missed it before - just fixed now.

File: pymc3/distributions/multivariate.py [hunk 6]
```diff
@@ -113,7 +124,10 @@ def logp(self, value):
         delta = value - mu
         k = tau.shape[0]
 
-        result = k * tt.log(2 * np.pi) + tt.log(1. / det(tau))
+        if self.gpu_compat is False:
```
Comment (twiecki) [inline pymc3/distributions/multivariate.py]: can just do `if self.gpu_compat: ... else: ...`

**Comment by twiecki** [comment]
Merging once tests pass.

**Comment by bhargavvader** [comment]
👍

**Comment by twiecki** [comment]
@bhargavvader I just realized that I gave bad advice with placing `logdet` into `theanof`, it should really be in `math.py`. Would it be too much to ask to change that?

File: pymc3/distributions/multivariate.py [hunk 7]
```diff
@@ -123,7 +124,10 @@ def logp(self, value):
         delta = value - mu
         k = tau.shape[0]
 
-        result = k * tt.log(2 * np.pi) - logdet(tau)
+        if gpu_compat is False:
+            result = k * tt.log(2 * np.pi) - logdet(tau)
+        elif gpu_compat is True:
+            result = k * tt.log(2 * np.pi) + tt.log(1. / det(tau))
```
Comment (ferrine) [inline pymc3/distributions/multivariate.py]: `result = k * tt.log(2 * np.pi) - tt.log(det(tau))` maybe

Comment (twiecki) [reply pymc3/distributions/multivariate.py]: or maybe even move the previous part on top and only do `result -= logdet(tau)`

File: pymc3/theanof.py [hunk 8]
```diff
@@ -59,11 +61,44 @@ def floatX(X):
...
+
+    Note: Once PR #3959 (https://github.com/Theano/Theano/pull/3959/) by harpone is merged,
+    this must be removed. 
+    """
+    def make_node(self, x):
+        x = theano.tensor.as_tensor_variable(x)
+        o = theano.tensor.scalar(dtype=x.dtype)
+        return Apply(self, [x], [o])
+
+    def perform(self, node, inputs, outputs):
```
Comment (ferrine) [inline pymc3/theanof.py:77]: Signature of overriden method changed, should be `perform(self, node, inputs, output_storage, params=None)` for consistency

Comment (ferrine) [reply pymc3/theanof.py:77]: That's not crucial thing, so you can take it in account only if you decide moving `logdet` to `math.py`

**Comment by bhargavvader** [comment]
I wouldn't mind changing it over to `math.py.` - should I open a different PR, or will you undo this one? I also wanted to know if there would be interest for a non-GPU compatible decorator like @fonnesbeck was talking about - in that case I can open the 2 PRs in a few days.

**Comment by twiecki** [comment]
Great! Yes, separate PR, also see @ferrine's point. if you have an idea for
the decorator sure but I don't think it's high priority.

File: pymc3/distributions/multivariate.py [hunk 9]
```diff
@@ -113,7 +124,10 @@ def logp(self, value):
         delta = value - mu
         k = tau.shape[0]
 
-        result = k * tt.log(2 * np.pi) + tt.log(1. / det(tau))
+        if self.gpu_compat:
+            result = k * tt.log(2 * np.pi) + tt.log(1. / det(tau))
```
Comment (twiecki) [inline pymc3/distributions/multivariate.py:128]: the first term can be moved up top, then the if-clause just does -logdet or -tt.log(det(tau)

**Comment by twiecki** [comment]
Seems like travis is unhappy about this all of a sudden:

```
======================================================================
ERROR: test_basic (pymc3.tests.test_math.TestLogDet)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/theano/gof/op.py", line 614, in __call__
    storage_map[ins] = [self._get_test_value(ins)]
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/theano/gof/op.py", line 570, in _get_test_value
    raise AttributeError('%s has no test value %s' % (v, detailed_err_msg))
AttributeError: <TensorType(float64, matrix)> has no test value  
Backtrace when that variable is created:
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/nose/suite.py", line 224, in run
    test(orig)
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/nose/case.py", line 45, in __call__
    return self.run(*arg, **kwarg)
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/nose/case.py", line 133, in run
    self.runTest(result)
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/nose/case.py", line 151, in runTest
    test(result)
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/unittest/case.py", line 649, in __call__
    return self.run(*args, **kwds)
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/unittest/case.py", line 601, in run
    testMethod()
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_math.py", line 31, in test_basic
    self.validate(test_case_1.astype(theano.config.floatX))
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_math.py", line 15, in validate
    x = theano.tensor.matrix()
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_math.py", line 31, in test_basic
    self.validate(test_case_1.astype(theano.config.floatX))
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_math.py", line 16, in validate
    f = theano.function([x], self.op(x))
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/theano/gof/op.py", line 628, in __call__
    (i, ins, node, detailed_err_msg))
ValueError: Cannot compute test value: input 0 (<TensorType(float64, matrix)>) of Op LogDet(<TensorType(float64, matrix)>) missing default value.  
Backtrace when that variable is created:
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/nose/suite.py", line 224, in run
    test(orig)
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/nose/case.py", line 45, in __call__
    return self.run(*arg, **kwarg)
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/nose/case.py", line 133, in run
    self.runTest(result)
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/nose/case.py", line 151, in runTest
    test(result)
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/unittest/case.py", line 649, in __call__
    return self.run(*args, **kwds)
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/unittest/case.py", line 601, in run
    testMethod()
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_math.py", line 31, in test_basic
    self.validate(test_case_1.astype(theano.config.floatX))
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_math.py", line 15, in validate
    x = theano.tensor.matrix()
```

Can you take a look @bhargavvader

File: pymc3/distributions/multivariate.py (hunk 9 above)
Comment (ferrine) [reply pymc3/distributions/multivariate.py:128]: You can use [this code](https://github.com/pymc-devs/pymc3/pull/1694/files#diff-e87226e0c9d21354a0fdf3a0e166ea18R161)  in future, it worked for me

**Comment by ferrine** [comment]
```python
x = tt.matrix()
x ** 2
```
fails with the same error
```python
x = tt.matrix()
x.tag.test_value = np.array([[1,1], [2,2]])
x ** 2
```
not

**Comment by bhargavvader** [comment]
Uh oh. I'll have a look at this. 
@ferrine , does it not fail when you set a test value? The error does mention it's missing a default.

**Comment by bhargavvader** [comment]
I'm a little confused about which `test_value` to use, and what the significance of it is; this [link](http://deeplearning.net/software/theano/tutorial/debug_faq.html) talks about how it's used to help debug, but I don't see how it's relevant here.  
When I try running the test after setting a random `(2,2)` matrix to test_value I get this error ` ValueError: Cannot compute test value: input 0 (input 0) of Op LogDet(input 0) missing default value.  `

This is similar to an [issue](https://github.com/Theano/Theano/issues/1439) raised by @fonnesbeck , where the solution is just to delete the cache (`.theano` directory). Do you think it might be a similar issue? Considering how it was fine before but failing now?

**Comment by bhargavvader** [comment]
@twiecki , what's the status of this? Is it still failing?

**Comment by twiecki** [comment]
Seems OK, not sure what caused this.

**Pull request merged**